### PR TITLE
Change taglist_get_tagstring to support 'unlimited' tag list size

### PR DESCRIPTION
**Pull request by dje29**
Previous taglist_get_tagstring signature/implementation did not allow
handling of cases where inputted buffer could not contain all tags.
New implementation allocates buffer based on pre-computed size allowing to
insert all tags in the returned string.

Added get_taglist_string in qthelper to handle conversion to QString
Added TestTagList with tests for taglist_get_tagstring

Signed-off-by: Jeremie Guichard <djebrest@gmail.com>

### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

This is a follow up on our discussion regarding this issue in #1184 when user would use so many tags that generated string would not fit in the 1024 buffer. I have (more or less) followed @neolit123 implementation proposal. 

### Changes made:

I've described all changes in the commit message.

Since I do not like to write code without tests, I've added some too.

### Release note:

Note sure if this change needs a mention in the CHANGELOG.md plus I would not know what topic to head it with, is there a place where a list of topics to be used here is maintained?

### Mentions:

@neolit123

### CODE REVIEW

File: core/dive.c
Comment (neolit123) [inline core/dive.c]: `char *tag...` rather than `char* tag...`

File: core/dive.c
Comment (bstoeger) [inline core/dive.c]: Please don't make it complicated for the caller (having to handle null). If there are no tags, return an empty string. Also makes the code here more straight-forward.

File: core/dive.c
Comment (neolit123) [inline core/dive.c]: you can just return `buffer` (`NULL`) here:
```
if (!buffer_size)
	return buffer;
```

File: core/dive.c
Comment (neolit123) [inline core/dive.c]:
```
if (!buffer) {
	fprintf(stderr, "taglist_get_tagstring(): cannot allocate buffer of size: u%\n);
	return NULL;
}
```
this would extract the contents of the branch bellow.

File: core/dive.c
Comment (neolit123) [inline core/dive.c]: space after `while`

File: core/dive.h
Comment (neolit123) [inline core/dive.h]: `char *`

File: tests/testtaglist.h
Comment (neolit123) [inline tests/testtaglist.h]: add space before `{`

File: tests/testtaglist.cpp
Comment (neolit123) [inline tests/testtaglist.cpp]: there is no let's change this to:
```
QCOMPARE(QString::fromUtf8(tagstring),
         QString::fromUtf8("A new tag, "
                           "A new tag 1, "
                           "A new tag 2, "
                           "A new tag 3, "
                           "A new tag 4, "
                           "A new tag 5")
);
```
notice:
- `, ` are at the end of the strings
- only spaces for the intent as this makes the indent consistent across tab sizes

File: core/dive.c
Comment (neolit123) [reply core/dive.c]: i think it's OK to return `NULL` as this is also covered in the tests.

**Review by neolit123 — CHANGES_REQUESTED** [review]
thanks for this. i've made some small comments about styling and the logic. please, amend.

extra points for including a test!

**Comment by neolit123** [comment]
> Note sure if this change needs a mention in the CHANGELOG.md plus I would not know what topic to head it with, is there a place where a list of topics to be used here is maintained?

there is no need for adding change log entries for internal changes like this one - that do not affect the user directly (e.g. code cleanup).

we don't really have a list of topics, but for this one something like `Dive-tags:` is a good one.

File: core/dive.c
Comment (neolit123) [inline core/dive.c]: let's use `unsigned int` here instead of `size_t`, this will save a cast when doing the `fprintf` later on.

File: tests/testtaglist.cpp
Comment (neolit123) [inline tests/testtaglist.cpp:2]: i would have moved the test into a separate commit, but i guess this is fine too. see if you'd prefer that?
the first commit message would need an amend in that case.

File: core/dive.c
Comment (bstoeger) [reply core/dive.c]: I think it's a really bad interface. The next person does a very reasonable
```
char *s = taglist_get_tagstring(...);
snprintf(buf, sizeof(buf), "Tags: %s\n", s);
free(s);
```
And boom  - strange and undefined behavior.

This code makes it more complicated for the caller *and* the callee. What's the advantage?

**Comment by bstoeger** [comment]
Wouldn't `membuffer` with `put_format()` lead to much simpler code with a negligible performance impact?

**Comment by neolit123** [comment]
it is possible to do it with `membuffer` but i dislike the fact that `realloc` will be called here for each tag entry.

**Comment by bstoeger** [comment]
> it is possible to do it with membuffer but i dislike the fact that realloc will be called here for each tag entry.

It won't:
```
                /* round it up to not reallocate all the time.. */
                needed = needed * 9 / 8 + 1024;
                n = realloc(b->buffer, needed);
```
In the normal case (< 1024 bytes) you'll get only a single allocation.

File: core/dive.c
Comment (neolit123) [reply core/dive.c]: i guess we can do that. please return `""` instead of `NULL`.
but this will also affect the test.

File: core/dive.c
Comment (neolit123) [inline core/dive.c]: this needs a small change. notice if a tag length is returned as `0` the size will still increment by `2`.
to avoid this store the old size and compare it instead of comparing to `0`.

File: core/dive.c
Comment (neolit123) [inline core/dive.c]: skip adding `, ` for empty tags?

**Comment by neolit123** [comment]
> It won't:
> In the normal case (< 1024 bytes) you'll get only a single allocation.

 i guess that's true. i leave it to @dje29 to decide if he wants to use `membuffer` with `put_format`.
examples in `core/save-xml.c`.

File: core/dive.c
Comment (bstoeger) [reply core/dive.c]: My idea was to remove both `if`-conditions - if there are not tags it would simply fall back to an empty string, no?

Comment (neolit123) [reply core/dive.c]: let's see the next iteration that @dje29 comes up with.

**Comment by dirkhh** [comment]
I wonder why the test failed - that needs to be investigated.
Great reviews, thanks @neolit123 and @bstoeger 
I agree that I'd love to see this use a membuffer. That would result in easier, cleaner code, IMHO.
This is always the balance of being welcoming and supportive to new contributors, and pushing for consistency and adherence to our standards... @dje29 would you be willing to change the implementation to use that instead?

**Comment by dje29** [comment]
Here is the change with new membuffer based implementation, I did not know this utility, it is quite handy and indeed made the code much simpler. Happy to have learned about it.

- As proposed by @bstoeger now function returns null terminated empty string when no tags are passed, tests is updated accordingly
- Fixed the issue raised by @neolit123 regarding printing ", " when list contains empty tag (added test cases for that)
- I have fixed coding style mistakes and hopefully did not introduce new ones... 

Working on this contribution made me think that CONTRIBUTING.md could be extended adding
1.  Section with some coding guidelines (such has use membuffer instead of malloc, strcpy, ... when working with strings), but certainly there are others that currently resid only in your heads and can only be discovered "the hard way" :) It would simply need the habit to patch the section every time a review like the one we just did is made. I'll prepare a pull request with a proposal for such entry
2. Mention of the CHANGELOG.md use (with list of topix that can be used) since currently new contributors discover this only once they create a pull request.

File: core/dive.c
Comment (bstoeger) [inline core/dive.c]: You can replace this `if`-block by
```
mb_cstring(&b);
```
I think the membuffer interface was  intended that way anyway.

File: core/dive.c
Comment (bstoeger) [inline core/dive.c]: You could write this as
```
if (!empty_string(tmp->tag->name))
```
This handles `name` being the null pointer gracefully.

**Comment by dirkhh** [comment]
@dje29 thanks for the rewrite.
I'd love a PR against CONTRIBUTING.md - I'm always happy to make things easier for new contributors joining our team.
As I mentioned on the mailing list, I'm planning to cut a release today, so we have a day or two to maybe consider the review by @bstoeger  :-)

**Comment by dje29** [comment]
@dirkhh I preferred to fix the code as the topic was still fresh in my head, but it is indeed a good idea to stall the merge to after the current release will be done.

**Comment by dirkhh** [comment]
Very nice - I like it. And the code is so much easier to read.
Thanks @dje29

**Pull request merged**
